### PR TITLE
Add threadpool impl and ut

### PR DESCRIPTION
**Pull request by M-iceberg**

### CODE REVIEW

File: linux/devdoc/threadpool_linux_requirements.md [hunk 1]
```diff
@@ -84,27 +84,27 @@ MOCKABLE_FUNCTION(, THREADPOOL_HANDLE, threadpool_create, EXECUTION_ENGINE_HANDL
...
+**SRS_THREADPOOL_LINUX_07_007: [** `threadpool_create` shall initialize every task item in the tasks array with `task_func` and `task_param` set to `NULL` and `task_state` set to `TASK_NOT_USED`. **]**
 
-`threadpool_create` shall create a SRW lock by calling `srw_lock_create`.
+**SRS_THREADPOOL_LINUX_07_008: [** `threadpool_create` shall create a SRW lock by calling `srw_lock_create`. **]**
 
-`threadpool_create` shall create a shared semaphore with initialized value zero.
+**SRS_THREADPOOL_LINUX_07_009: [** `threadpool_create` shall create a shared semaphore with initialized value zero. **]**
 
-`insert_idx`  and `consume_idx` for the task array shall be initialized to 0.
+**SRS_THREADPOOL_LINUX_07_010: [** `insert_idx`  and `consume_idx` for the task array shall be initialized to 0. **]**
```
Comment (dcristoloveanu) [inline linux/devdoc/threadpool_linux_requirements.md:105]: >   [](http://example.com/codeflow?start=47&length=2)

Nitpick: extra space

File: linux/src/threadpool_linux.c [hunk 2]
```diff
@@ -312,28 +246,17 @@ static int reallocate_threadpool_array(THREADPOOL* threadpool)
...
-                // Consume = 3
-                // Produce = 2
-                // [x x 0 x]
-                // During resize we will have to memmove in the gap at 2
-                // [x x 0 x 0 0 0 0]
-
                 int64_t insert_pos = interlocked_add_64(&threadpool->insert_idx, 0) % existing_count;
                 int64_t consume_pos = interlocked_add_64(&threadpool->consume_idx, 0) % existing_count;
-
                 uint32_t compress_count = 0;
```
Comment (dcristoloveanu) [inline linux/src/threadpool_linux.c:251]: >compress [](http://example.com/codeflow?start=25&length=8)

It's not really compress as in compression (deflate, zip, etc.)  Likely a better name would be move_count

File: linux/src/threadpool_linux.c [hunk 3]
```diff
@@ -312,28 +246,17 @@ static int reallocate_threadpool_array(THREADPOOL* threadpool)
...
+
+                /* Codes_SRS_THREADPOOL_LINUX_07_044: [ threadpool_schedule_work shall remove any gap in the task array. ]*/
```
Comment (dcristoloveanu) [inline linux/src/threadpool_linux.c:253]: >all remove any gap [](http://example.com/codeflow?start=83&length=18)

I think a simpler way to do this is to memmove everything between the consume index and the size of the array before resize (so one element in the example above) to the end of the new resized array.

No iteration one by one, no gap filling, simply moving a bunch of entries.

Comment (dcristoloveanu) [reply linux/src/threadpool_linux.c:253]: Let me know if it does not make sense and we can whiteboard it.

**Review by dcristoloveanu — DISMISSED** [review]
:shipit:

File: linux/devdoc/threadpool_linux_requirements.md (hunk 1 above)
Comment (M-iceberg) [reply linux/devdoc/threadpool_linux_requirements.md:105]: deleted

File: linux/src/threadpool_linux.c (hunk 3 above)
Comment (M-iceberg) [reply linux/src/threadpool_linux.c:253]: done

File: linux/src/threadpool_linux.c [hunk 4]
```diff
@@ -312,28 +246,17 @@ static int reallocate_threadpool_array(THREADPOOL* threadpool)
...
 
-                // Ensure there are no gaps in the array
-                // Consume = 2
-                // Produce = 2
-                // [x x x x]
-                // Consume = 3
-                // Produce = 2
-                // [x x 0 x]
-                // During resize we will have to memmove in the gap at 2
-                // [x x 0 x 0 0 0 0]
```
[the remark this answers is not in the document]
Comment (M-iceberg) [reply linux/src/threadpool_linux.c:323]: added back

File: linux/src/threadpool_linux.c (hunk 2 above)
Comment (M-iceberg) [reply linux/src/threadpool_linux.c:251]: deleted

**Comment by M-iceberg** [comment]
>         LogError("Invalid Argument timer_instance");

done

---
In reply to: [1461472109](https://github.com/Azure/c-pal/pull/223#issuecomment-1461472109) [](http://example.com/codeflow?ancestors=1461472109,1461472109)

---
Refers to: linux/src/threadpool_linux.c:99 in 2738d00. [](commit_id = 2738d00f13192e4d6c51fffc4bea78a43d9528bf, deletion_comment = False)

**Review by dcristoloveanu — APPROVED** [review]
:shipit:

**Pull request merged**